modules/parachains/src/lib.rs

```diff
@@ -337,112 +333,113 @@ pub mod pallet {
+			let mut storage = GrandpaPallet::<T, I>::storage_proof_checker(
 				relay_block_hash,
 				parachain_heads_proof.0,
-				move |mut storage| {
-					for (parachain, parachain_head_hash) in parachains {
-						let parachain_head = match Pallet::<T, I>::read_parachain_head(&mut storage, parachain) {
-							Ok(Some(parachain_head)) => parachain_head,
-							Ok(None) => {
-								log::trace!(
-									target: LOG_TARGET,
-									"The head of parachain {:?} is None. {}",
-									parachain,
-									if ParasInfo::<T, I>::contains_key(parachain) {
-										"Looks like it is not yet registered at the source relay chain"
-									} else {
-										"Looks like it has been deregistered from the source relay chain"
-									},
-								);
-								Self::deposit_event(Event::MissingParachainHead { parachain });
-								continue;
-							},
-							Err(e) => {
-								log::trace!(
-									target: LOG_TARGET,
-									"The read of head of parachain {:?} has failed: {:?}",
-									parachain,
-									e,
-								);
-								Self::deposit_event(Event::MissingParachainHead { parachain });
-								continue;
-							},
-						};
-
-						// if relayer has specified invalid parachain head hash, ignore the head
-						// (this isn't strictly necessary, but better safe than sorry)
-						let actual_parachain_head_hash = parachain_head.hash();
-						if parachain_head_hash != actual_parachain_head_hash {
+			)
+			.map_err(Error::<T, I>::HeaderChainStorageProof)?;
+
+			for (parachain, parachain_head_hash) in parachains {
+				let parachain_head =
+					match Pallet::<T, I>::read_parachain_head(&mut storage, parachain) {
+						Ok(Some(parachain_head)) => parachain_head,
+						Ok(None) => {
 							log::trace!(
 								target: LOG_TARGET,
-								"The submitter has specified invalid parachain {:?} head hash: {:?} vs {:?}",
+								"The head of parachain {:?} is None. {}",
 								parachain,
-								parachain_head_hash,
-								actual_parachain_head_hash,
+								if ParasInfo::<T, I>::contains_key(parachain) {
+									"Looks like it is not yet registered at the source relay chain"
+								} else {
+									"Looks like it has been deregistered from the source relay chain"
+								},
 							);
-							Self::deposit_event(Event::IncorrectParachainHeadHash {
+							Self::deposit_event(Event::MissingParachainHead { parachain });
+							continue
+						},
+						Err(e) => {
+							log::trace!(
+								target: LOG_TARGET,
+								"The read of head of parachain {:?} has failed: {:?}",
+								parachain,
+								e,
+							);
+							Self::deposit_event(Event::MissingParachainHead { parachain });
+							continue
+						},
+					};
+
+				// if relayer has specified invalid parachain head hash, ignore the head
+				// (this isn't strictly necessary, but better safe than sorry)
+				let actual_parachain_head_hash = parachain_head.hash();
+				if parachain_head_hash != actual_parachain_head_hash {
+					log::trace!(
```

Comment:
looks like formatting is no good for this trace

Reply:
You're right. I didn't notice it. Fixed.